cmd/silkworm.cpp

```diff
@@ -214,37 +215,85 @@ int main(int argc, char* argv[]) {
+                auto genesis_json = nlohmann::json::parse(source_data, nullptr, /* allow_exceptions = */ false);
+                if (genesis_json.is_discarded()) {
+                    throw std::runtime_error("Could not initialize db for chain id " +
+                                             std::to_string(node_settings.network_id) + " : unknown network");
+                }
+                log::Message() << "Initializing db for chain_id " << node_settings.network_id;
+                db::initialize_genesis(tx, genesis_json, /*allow_exceptions=*/true);
+                tx.commit();
+                tx = chaindata_env.start_write();
+                db_chain_config = db::read_chain_config(tx);
+            }
+
+            log::Message() << "Chain configuration " << db_chain_config.value().to_json().dump();
+            if (db_chain_config.value().chain_id != node_settings.network_id) {
+                throw std::runtime_error("Network Id incompatible. Expected " +
+                                         std::to_string(node_settings.network_id) + " got " +
+                                         std::to_string(db_chain_config.value().chain_id));
+            }
+        }
+
+        // Detect prune-mode and verify is compatible
+        {
+            auto tx{chaindata_env.start_write()};
+            auto db_prune_mode{db::read_prune_mode(tx)};
+            if (db_prune_mode != node_settings.prune_mode) {
+                // In case we have mismatching modes (cli != db) we prevent
+                // further execution ONLY if we've already synced something
+                auto header_download_progress{db::stages::read_stage_progress(tx, db::stages::kHeadersKey)};
+                if (header_download_progress) {
+                    throw std::runtime_error("Can't change prune_mode on already synced data. Expected " +
+                                             node_settings.prune_mode.to_string() + " got " +
+                                             db_prune_mode.to_string());
+                }
+                db::write_prune_mode(tx, node_settings.prune_mode);
+                tx.commit();
+                tx = chaindata_env.start_read();
+                db_prune_mode = db::read_prune_mode(tx);
+            }
+            log::Message() << "Prune mode " << db_prune_mode.to_string();
+        }
+
+        // Do sync stuff here
+
+        log::Message() << "Closing Database chaindata path " << node_settings.data_directory->chaindata().path();
         chaindata_env.close();
 
     } catch (const CLI::ParseError& ex) {
         return cli.exit(ex);
+    } catch (const std::runtime_error& ex) {
+        log::Error() << ex.what();
+        return -1;
     } catch (const std::invalid_argument& ex) {
-        std::cerr << "Invalid argument :" << ex.what() << "\n" << std::endl;
+        std::cerr << "\tInvalid argument :" << ex.what() << "\n" << std::endl;
```

Comment:
Why not `log::Error` ?

Reply:
Cause this catch is related to the `try` block where log initialization occurs. We might not have it ready yet.